sympy/solvers/ode.py

```diff
@@ -2870,8 +2883,11 @@ def _conditional_term_factoring(expr):
     return expr
 
 
-def constant_renumber(expr, symbolname, startnumber, endnumber):
+def constant_renumber(eq, expr):
```

Comment:
What about making the signature `(expr, exclude=None)`? The user would then pass `(expr, eq.free_symbols)`.

Is this just an internal function that we are free to change?

Reply:
The docstring for the ode module says that these are internal functions:
[https://github.com/sympy/sympy/blob/2d4eb2fd9f8333bb1992c31069f256db542b589b/sympy/solvers/ode.py#L30](https://github.com/sympy/sympy/blob/2d4eb2fd9f8333bb1992c31069f256db542b589b/sympy/solvers/ode.py#L30)

We could do have `exclude=None` but the signature would need to be:
```python
def constant_renumber(expr, constants, exclude):
```
so that we know which symbols in expr are constants *and* which symbols not in expr would still clash with the ODE.

Reply:
> docstring for the ode module says that these are internal functions

Then I think we are ok changing them. The approach here seems much less fragile and more user friendly.

I'm not sure I understand the subtlety of the distinction between `exclude` and `foo.free_symbols`. It seems to me that `eq` is included with `expr` (in `(eq, expr)`) so you can identify which constants were introduced into `expr` that were not in `eq`...I am just suggesting that you let the user tell you what symbols to ignore with `exclude` instead of passing `eq`.

Reply:
There is no subtlety in the distinction. At the moment this PR allows you to do:
```python
sol_renumbered = constant_renumber(eq_ode, sol)
```
If we changed it to have `(expr, constants, exclude)` as arguments then the equivalent would be:
```python
sol_renumbered = constant_renumber(sol,
     constants=sol.free_symbols - eq_ode.free_symbols,
     exclude=eq_ode.free_symbols)
```
That might seem cumbersome for the caller but this function is called twice in the whole module (including once recursively) so it's not a big deal for the caller to prepare these arguments. The tests call this function a lot but we could make a helper there.

So how about:
```python
def constant_renumber(expr, constants, exclude=None):
```
where exclude is an optional list (not needed in the tests) and constants is a list of symbols.

Then in the tests we can do:
```diff
-    sol8s = constant_renumber(sol8, 'C', 1, 4)
+    sol8s = constant_renumber(sol8, symbols('C1:5'))
```

The change to `odesimp` can then just be a new optional `exclude` argument.

Reply:
Another possibility (bearing in mind #11024) is that the signature can be:
```python
def constant_renumber(expr, old_constants, new_constants_iter=None):
```
Then inside `dsolve` we can pass `new_constants_iter=iter_numbered_symbols(eq_ode, start=1, prefix='C')` which will generate new constants not found in the ODE. We can then document that users calling `dsolve` with their own constants iterator will need to ensure that it only generates symbols not found in the ODE (e.g. by using `iter_numbered_symbols`).

Yet another possibility would be
```python
def constant_renumber(expr, old_constants, exclude=None, new_constants_iter=None):
```
and `constant_renumber` can do the checking to ensure that constants coming from `new_constants_iter` don't clash with those in `exclude`. That would be slightly friendlier on `dsolve` users who want to pass in their own constant iterator.

Reply:
I would make it
```python
def constant_renumber(sol, exclude=None, cons=None):
    exclude = set(exclude or [])
    constants=sol.free_symbols - set(exclude or [])
    ...
```
And call it as
```python
constant_renumber(sol, eq_ode.free_symbols, numbered_constants('C'))
```
> need to ensure that it only generates symbols not found in the ODE

I would just keep generating symbols until no clashes are found. So the user basically decides on what symbol name will be used and numbered symbols are supplied that are not already in use.

Reply:
> I would make it
> 
> ```python
> def constant_renumber(sol, exclude=None, cons=None):
> ```

Yeah, that looks good. Maybe exclude should be called `nonconstants` or something.

Reply:
> should be called `nonconstants` or something.

or `variables`

Will `constant_renumber` only target numbered constants or would it use numbered constants for any constants that weren't excluded? e.g. 
```python
>> constant_renumber(x + y, exclude=y, numbered_symbols('c'))
c0 + y
```

Reply:
With the signature that you have just proposed any symbol found in sol but not in exclude will be replaced with a symbol from cons (which will default to C1, C2, ...) so that would be:
```python
>>> constant_renumber(x + y, exclude=[y], iter_numbered_symbols(eq_ode, 'c'))
c1 + y
```

I think `variables` here is not quite right. In `Eq(f(x).diff(x), a*x)` we would normally refer to `f` and `x` as the dependent and independent variables. The symbol `a` is a constant but not an integration constant. We need `exclude` to be the list of all symbols in the solution that are not integration constants.

We're probably overthinking this though since this is an internal function. I will go make the changes.

Reply:
I guess I am thinking of something more general: a function that replaces constants with numbered symbols that don't clash with any variables in the original expression. That is why I use `numered_constants('c')` rather than refer to `eq_ode` in `iter_numered_symbols`. So the function basically needs to know the original expression and which symbols *not* to target and then just numbers everything else, making sure not to clash with an existing symbol.
```python
def syms_numbered(current, fixed, syms):
    old = list(ordered(set(current) - set(fixed)))
    new = []
    for i in old:
        while True:
            s = next(syms)
            if s not in fixed:
                new.append(s)
                break
    return dict(zip(old, new))  # use as eq.xreplace(rv)
```
e.g.
```python
>>> syms_numbered([x,y,var('c1')],[y],numbered_symbols('x'))
{c1: x0, x: x1}
```

Reply:
That's more or less where I ended up yesterday. I didn't push at the time because of the build queue. I'll push it now.